llama_index/embeddings/clarifai.py

```diff
@@ -0,0 +1,135 @@
+import logging
+from typing import Any, List, Optional
+
+from llama_index.bridge.pydantic import Field, PrivateAttr
+from llama_index.callbacks import CallbackManager
+from llama_index.embeddings.base import DEFAULT_EMBED_BATCH_SIZE, BaseEmbedding
+
+logger = logging.getLogger(__name__)
+
+EXAMPLE_URL = "https://clarifai.com/anthropic/completion/models/claude-v2"
+
+
+class ClarifaiEmbedding(BaseEmbedding):
+    """Clarifai embeddings class.
+
+    Clarifai uses Personal Access Tokens(PAT) to validate requests.
+    You can create and manage PATs under your Clarifai account security settings.
+    Export your PAT as an environment variable by running `export CLARIFAI_PAT={PAT}`
+    """
+
+    model_url: Optional[str] = Field(
+        description=f"Full URL of the model. e.g. `{EXAMPLE_URL}`"
+    )
+    model_id: Optional[str] = Field(description="Model ID.")
+    model_version_id: Optional[str] = Field(description="Model Version ID.")
+    app_id: Optional[str] = Field(description="Clarifai application ID of the model.")
+    user_id: Optional[str] = Field(description="Clarifai user ID of the model.")
+
+    _model: Any = PrivateAttr()
+
+    def __init__(
+        self,
+        model_name: Optional[str] = "",
+        model_url: Optional[str] = "",
+        model_version_id: Optional[str] = "",
+        app_id: Optional[str] = None,
+        user_id: Optional[str] = None,
+        embed_batch_size: int = DEFAULT_EMBED_BATCH_SIZE,
+        callback_manager: Optional[CallbackManager] = None,
+    ):
+        try:
+            from clarifai.client.model import Model
+        except ImportError:
+            raise ImportError("ClarifaiEmbedding requires `pip install clarifai`.")
+
+        if model_url != "" and model_name != "":
+            raise ValueError("You can only specify one of model_url or model_name.")
+        if model_url == "" and model_name == "":
+            raise ValueError("You must specify one of model_url or model_name.")
+
+        if model_url != "":
+            self._model = Model(model_url)
+
+        if model_name != "":
+            if app_id == "" or user_id == "":
```

Comment:
very very nit: edge cases when users set `model_url` and/or `model_name` as `None`.

Would be to explicitly type as non-nullable, or properly handle either `None` or empty strings.